onnxruntime/core/providers/webnn/builders/impl/gqa_op_builder.cc

```diff
@@ -0,0 +1,558 @@
+  }
+
+  // Reshape and transpose the input "query"
+  common_options.set("label", node.Name() + "_/GQA/query/reshape");
+  emscripten::val reshaped_query = model_builder.GetBuilder().call<emscripten::val>(
+      "reshape", query_input, emscripten::val::array(reshape_tensor_shape), common_options);
+
+  emscripten::val transpose_options = emscripten::val::object();
+  transpose_options.set("permutation", emscripten::val::array(std::vector<uint32_t>({0, 2, 1, 3})));
+  transpose_options.set("label", node.Name() + "_/GQA/query/transpose");
+  emscripten::val new_query =
+      model_builder.GetBuilder().call<emscripten::val>("transpose", reshaped_query, transpose_options);
+
+  // Reshape the inputs "key" and "value" for scatterND
+  std::vector<uint32_t> reshape_kv_shape = {batch_size, qkv_sequence_length, kv_num_heads, head_size};
+  common_options.set("label", node.Name() + "_/GQA/key/reshape_1");
+  emscripten::val key_for_scatter = model_builder.GetBuilder().call<emscripten::val>(
+      "reshape", key_input, emscripten::val::array(reshape_kv_shape), common_options);
+
+  common_options.set("label", node.Name() + "_/GQA/value/reshape_1");
+  emscripten::val value_for_scatter = model_builder.GetBuilder().call<emscripten::val>(
+      "reshape", value_input, emscripten::val::array(reshape_kv_shape), common_options);
+
+  /* Calculate scatter_indices for kv's scatterND
+                                                                                 if_prefill (0/1 constant)
+                                                                                          |
+  scatter_indices_left_constant         scatter_indices_right_constant           0 ---> Where <--- Cast <---seqlens_k
+                |                                      |                                  |
+                |                                     Add <--------------------------- scatter_pos*
+                |                                      |
+                +------------------+-------------------+
+                                   |
+                            scatter_indices
+  */
+  // Prepare the constant materials for scatter indices
+  auto left = generate_indices(batch_size, kv_num_heads, qkv_sequence_length);
+  auto right = repeat_sequence(qkv_sequence_length, kv_num_heads, batch_size);
+
+  ORT_RETURN_IF_NOT(SetWebnnDataType(common_desc, ONNX_NAMESPACE::TensorProto_DataType_INT32), "Unsupported data type");
+  emscripten::val dims_left =
+      emscripten::val::array(std::vector<uint32_t>({batch_size * qkv_sequence_length * kv_num_heads, 2}));
+  common_desc.set("dimensions", dims_left);
+  common_desc.set("shape", dims_left);
+  emscripten::val left_buffer = emscripten::val::global("Int32Array").new_(emscripten::val::array(left));
+  emscripten::val left_constant =
+      model_builder.GetBuilder().call<emscripten::val>("constant", common_desc, left_buffer);
```

Comment:
Can we share the WebNN constant for each GQA to save some memory?

e.g. For a Phi-3-mini, it contains 32 GQA nodes, all these GQA nodes have the same batch_size, kv_num_heads, qkv_sequence_length..., so their `left_constant` and `right_constant` should be same.